compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -969,12 +969,38 @@ impl<'p, 'tcx> Usefulness<'p, 'tcx> {
                         // constructor, that matches everything that can be built with
                         // it. For example, if `ctor` is a `Constructor::Variant` for
                         // `Option::Some`, we get the pattern `Some(_)`.
-                        split_wildcard
-                            .iter_missing(pcx)
-                            .map(|missing_ctor| {
-                                Fields::wildcards(pcx, missing_ctor).apply(pcx, missing_ctor)
-                            })
-                            .collect()
+                        let mut new = vec![];
+                        for missing_ctor in split_wildcard.iter_missing(pcx) {
+                            if let Constructor::Variant(idx) = missing_ctor {
+                                if let ty::Adt(adt, _) = pcx.ty.kind() {
+                                    if pcx
+                                        .cx
+                                        .tcx
+                                        .get_attrs(adt.variants[*idx].def_id)
+                                        .iter()
+                                        .find_map(|attr| {
+                                            if attr.has_name(sym::doc) {
+                                                attr.meta_item_list()
+                                            } else {
+                                                None
+                                            }
+                                        })
+                                        .map(|items| {
+                                            items
+                                                .iter()
+                                                .find(|item| item.has_name(sym::hidden))
+                                                .is_some()
+                                        })
+                                        .unwrap_or_default()
+                                    {
+                                        new = vec![Pat::wildcard_from_ty(pcx.ty)];
```

Comment:
This logic won't do what we want for
```rust
match Foo::A {
	Foo::A => {}
}
```
We probably want to just skip the hidden variants here, and add a wildcard at the end if we skipped anything